src/lib/parser.rs

```diff
@@ -0,0 +1,217 @@
+    }
+
+    fn off_to_line_col(&self, off: usize) -> (usize, usize) {
+        if off == self.src.len() {
+            let line_off = *self.newlines.iter().last().unwrap();
+            return (self.newlines.len(), self.src[line_off..].chars().count() + 1);
+        }
+        let (line_m1, &line_off) = self.newlines.iter()
+                                                .enumerate()
+                                                .rev()
+                                                .find(|&(_, &line_off)| line_off <= off)
+                                                .unwrap();
+        let c_off = self.src[line_off..]
+                        .char_indices()
+                        .position(|(c_off, _)| c_off == off - line_off)
+                        .unwrap();
+        return (line_m1 + 1, c_off + 1);
+    }
+
+    fn parse(&mut self) -> LexBuildResult<usize> {
+        let mut i = try!(self.parse_declarations(0));
+        i = try!(self.parse_rules(i));
+        // We don't currently support the subroutines part of a specification. One day we might...
+        match self.lookahead_is("%%", i) {
+            Some(j) => {
+                if try!(self.parse_ws(j)) == self.src.len() { Ok(i) }
+                else {
+                    Err(self.mk_error(LexErrorKind::RoutinesNotSupported, i))
+                }
+            }
+            None    => Ok(i)
+        }
+    }
+
+    fn parse_declarations(&mut self, mut i: usize) -> LexBuildResult<usize> {
+        i = try!(self.parse_ws(i));
+        if self.lookahead_is("%%", i).is_some() { return Ok(i + 2); }
+        if i < self.src.len() {
+            Err(self.mk_error(LexErrorKind::UnknownDeclaration, i))
+        }
+        else {
+            Err(self.mk_error(LexErrorKind::PrematureEnd, i - 1))
+        }
+    }
+
+    fn parse_rules(&mut self, mut i: usize) -> LexBuildResult<usize> {
+        loop {
+            i = try!(self.parse_ws(i));
+            if i == self.src.len() { break; }
+            if self.lookahead_is("%%", i).is_some() { break; }
+            i = try!(self.parse_rule(i));
+        }
+        Ok(i)
+    }
+
+    fn parse_rule(&mut self, i: usize) -> LexBuildResult<usize> {
+        let line_len = self.src[i..]
+                       .chars()
+                       .take_while(|&c| c != '\n')
+                       .fold(0, |a, c| a + c.len_utf8());
+        let line     = self.src[i..i + line_len].trim_right();
+        let rspace   = match line.rfind(' ') {
+            Some(j) => j,
+            None    => return Err(self.mk_error(LexErrorKind::MissingSpace, i))
+        };
+
+        let name;
+        let orig_name = &line[rspace + 1..];
+        if orig_name == ";" {
+            name = None;
+        }
+        else if self.ast.get_rule(orig_name).is_some() {
+            return Err(self.mk_error(LexErrorKind::DuplicateName, i + rspace + 1))
+        }
+        else {
+            name = Some(orig_name.to_string());
```

Comment:
Doesn't this allow us to use anything as a token name? Maybe we should match `orig_name` to `[a-zA-Z_]`.